File: reactionver/ConnectBoxes.java
import java.io.File;
import java.io.IOException;
import java.text.DecimalFormat;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ConnectBoxes {
	
	Atom atomArray[] = null;
	double atomdistanArray[][];
	double maxdistan = Double.MIN_VALUE;
	double mindiatan = Double.MAX_VALUE;
	double lr;
	//cave radius
	double cr;
	double alpha;
	double gamma;
	double rmod;
	int gbmod;
	int pushmod;
	private Scanner sca;
	
	ConnectBoxes (String pdbpath, 
				  double prl, double prh, 
				  double gx, double gy, double gz, 
				  int boxnum,
				  double lr, 
				  double beta, double alpha, int rmodule, int gbmod, double mr, int pushmod) throws IOException, InterruptedException{
		//this is multiplier of lr to nebor distances
		cr = lr * beta;
		//this is multiplier of lr to push
		this.lr = lr;
		this.alpha = alpha;
		this.gamma = 0.1;
		this.rmod = rmodule;
		this.gbmod = gbmod;
		this.pushmod = pushmod;
		//getCA from pdbqt
		getCA(pdbpath);
		getDistanceArray();
		if(rmodule == 0){
			getlrNeborNum(cr);
		}
		else if(rmodule == 1){
			getBestSDneborNum();
		}
		else if(rmodule == 2){
			getlrNeborNum(mr);
		}
		else if(rmodule == 3){
			getBestSDupbound(mr);
		}
		Arrays.sort(atomArray);
		//search good sentinel boxnum times
			System.out.println("Blue Atom:");
		for(int i = (int)(prh*(atomArray.length)) ; i < atomArray.length ; i++)
		{
			System.out.print(atomArray[i].x);
			System.out.print(" " + atomArray[i].y);
			System.out.println(" " + atomArray[i].z);
		}
			System.out.println("Surface Atom:");
		for(int i = (int)(prl*(atomArray.length)) ; i < (prh*(atomArray.length)) ; i++)
		{
			System.out.print(atomArray[i].x);
			System.out.print(" " + atomArray[i].y);
			System.out.println(" " + atomArray[i].z);
		}
			System.out.println("First Box:");
		getSentinel(1, prl, prh, gx, gy, gz);
		//System.out.println(maxdistan);
		
	}
	
	public void getCA(String path) throws IOException, InterruptedException{
		Pattern pat;
        Matcher mat;
        Boolean found;
		File f = new File(path+".pdbqt");
		sca = new Scanner(f);
		Queue<String> q = new LinkedList<String>(); 
	    pat = Pattern.compile("ATOM +[0-9]+ +CA");
		while(sca.hasNext()){
			String t = sca.nextLine();
			mat = pat.matcher(t);
			found = mat.find();
			if(found){
				q.offer(t);
			}
		}
	    atomArray = new Atom[q.size()];
	    for(int i=0 ; i<atomArray.length ; i++){
        	atomArray[i] = new Atom();
        	String t[] = ((String)q.poll()).split(" +");
        	atomArray[i].serial = Integer.parseInt(t[1]);
        	atomArray[i].x = Double.parseDouble(t[6]);
        	atomArray[i].y = Double.parseDouble(t[7]);
        	atomArray[i].z = Double.parseDouble(t[8]);
        }
	}
	
	public void getSentinel(int boxnum, double prl, double prh , double gx, double gy, double gz){
		for(int i=0 ; i<boxnum ; i++){
			Atom ansAtom = getGoodSentinel((int)(prl*(atomArray.length)), getBlueAtom(prh, this.gbmod), gx, gy, gz);
			DecimalFormat df=new DecimalFormat("#.###");
			if(ansAtom==null){
				System.out.println("null");
				System.out.println("null");
				System.out.println("null");
				break;
			}
			else{
				System.out.print(df.format(ansAtom.x));
				System.out.print(" " + df.format(ansAtom.y));
				System.out.println(" " + df.format(ansAtom.z));
			}
		}
	}
	
	public Atom getGoodSentinel(int lindex, int hindex, double gx, double gy, double gz){
		int ans = lindex;
		//go to correct start atom
		while(ans<hindex){
			if(atomArray[ans].neborNum>0)
				break;
			ans++;
		}
		if(ans == hindex){
			return null;
		}
		//if i and j is nebor then compare i's neborNum with j
		for(int j=lindex ; j<hindex ; j++){
			double x = Math.abs(atomArray[ans].x-atomArray[j].x);
			double y = Math.abs(atomArray[ans].y-atomArray[j].y);
			double z = Math.abs(atomArray[ans].z-atomArray[j].z);
			if( x<gx && y<gy && z<gz && (atomArray[j].neborNum > atomArray[ans].neborNum) )
				ans = j;
		}
		deleteNeborNum(ans, lindex, hindex, gx, gy ,gz);
		return pushOutAns(ans, hindex);
	}
	
	public int getBlueAtom(double prh, int mod){
		if(mod == 0){
			//將範圍中，鄰居數最多者的鄰居數prh%的數量為基準，數量再其以上為藍點
			int fullnb = (int) (atomArray[atomArray.length-1].neborNum*prh);
			//System.out.println("full: "+fullnb);
			Atom t = new Atom();
			t.neborNum = fullnb;
			//System.out.println("Max: "+atomArray[atomArray.length-1].neborNum);
			int i = Arrays.binarySearch(atomArray, t);
			if(i < 0)
				i = Math.abs(i+1);
			return i;
		}
		else {
			return (int) (prh*(atomArray.length));
		}
	}
		
	public Atom pushOutAns (int ans, int hindex){
		Atom tempAtom = new Atom(atomArray[ans]);
		double[] uv= {0, 0, 0};
		//宣告blueArray，並初始化
		double[][] blueArray;
		//如果pushmod為0，代表不指定靠近的名次，則使用近gamma%的blueAtom
		if(pushmod==0){
			blueArray = new double[(int) Math.ceil(atomArray.length-hindex*gamma)][2];
		}
		//如果pushmod不為0，則使用近pushmod名的buleAtom
		else{
			blueArray = new double[pushmod][2];
		}
		for(int i=0 ; i<blueArray.length ; i++)
			blueArray[i][1]=Double.MAX_VALUE;
		//取得前gamma%或pushmod的blueAtom
		for(int i=hindex ; i<atomArray.length ; i++){
			double td = getDistance(atomArray[ans].x, atomArray[ans].y, atomArray[ans].z, atomArray[i].x, atomArray[i].y, atomArray[i].z);
			//將已有的blueAtom距離跟newd比對，較小則取代
			for(int j=0 ; j<blueArray.length ; j++){
				if(td < blueArray[j][1]){
					//紀錄atom索引，跟距離
					blueArray[j][0] = i;
					blueArray[j][1] = td;
					break;
				}
			}
		}
		//get blueAtom to ans Unit vector
		for(int i=0 ; i<blueArray.length ; i++){
			double x = atomArray[ans].x-atomArray[(int)blueArray[i][0]].x;
			double y = atomArray[ans].y-atomArray[(int)blueArray[i][0]].y;
			double z = atomArray[ans].z-atomArray[(int)blueArray[i][0]].z;
			//單位向量
			x = x/blueArray[i][1];
			y = y/blueArray[i][1];
			z = z/blueArray[i][1];
			//若pushmod為0，則加上權重
			if(pushmod==0){
				x = x/blueArray[i][1];
				y = y/blueArray[i][1];
				z = z/blueArray[i][1];
			}
			uv[0] = uv[0]+x;
			uv[1] = uv[1]+y;
			uv[2] = uv[2]+z;
		}
		
		double ud = Math.sqrt(Math.pow(uv[0], 2)+Math.pow(uv[1], 2)+Math.pow(uv[2], 2));
		//System.out.println(uv[0]+"-"+uv[1]+"-"+uv[2]);
		uv[0] = uv[0]/ud;
		uv[1] = uv[1]/ud;
		uv[2] = uv[2]/ud;
		//System.out.println(uv[0]+"-"+uv[1]+"-"+uv[2]);
		double a = lr*this.alpha;
		tempAtom.x += uv[0]*a;
		tempAtom.y += uv[1]*a;
		tempAtom.z += uv[2]*a;
		return tempAtom;
	}
	
	public void deleteNeborNum(int ans, int lindex, int hindex, double gx, double gy, double gz){
		for(int j=lindex ; j<hindex ; j++){
			double x = Math.abs(atomArray[ans].x-atomArray[j].x);
			double y = Math.abs(atomArray[ans].y-atomArray[j].y);
			double z = Math.abs(atomArray[ans].z-atomArray[j].z);
			//if i and j is nebor, delete atome
			if( x<gx && y<gy && z<gz /*&& i!=j*/){
				atomArray[j].neborNum = -1;
			}
		}
	}

	public void getDistanceArray(){
		atomdistanArray = new double[atomArray.length][atomArray.length];
		for(int i=0 ; i<atomArray.length ; i++){
			for (int j=i ; j<atomArray.length ; j++){
				atomdistanArray[i][j] = getDistance(atomArray[i].x, atomArray[i].y, atomArray[i].z, 
			               atomArray[j].x, atomArray[j].y, atomArray[j].z);
				atomdistanArray[j][i] = atomdistanArray[i][j];
				if(atomdistanArray[i][j] > maxdistan)
					maxdistan = atomdistanArray[i][j];
				if(atomdistanArray[i][j] < mindiatan && atomdistanArray[i][j] != 0)
					mindiatan = atomdistanArray[i][j];
			}
			Arrays.sort(atomdistanArray[i]);
		}
	}
	
	public void getBestSDneborNum(){
		//System.out.println(mindiatan+"-"+maxdistan);
		double SD=0;
		int continu=5; 
		double avgrArray[] = getAvgrArray(2, 1);
		for(int r=0 ; r<avgrArray.length && continu>0 ; r++){
			double avg=0;
			double sum1=0;
			double sum2=0;
			double newSD = 0;
			int tempNeborNum[] = new int[atomArray.length];
			getNeborNum(avgrArray[r], tempNeborNum);
			//compute SD
			for(int i=0 ; i<tempNeborNum.length ; i++){
				sum1 += tempNeborNum[i];
				sum2 += Math.pow(tempNeborNum[i], 2);
			}
			avg = sum1 / tempNeborNum.length;
			newSD = Math.sqrt( (sum2/tempNeborNum.length)-Math.pow(avg, 2) );
			//get better neborNum if SD is better
			if (newSD > SD){
				SD = newSD;
				continu = 5;
				for(int i=0 ; i<atomArray.length ; i++)
					atomArray[i].neborNum = tempNeborNum[i];
				this.cr = avgrArray[r];
			}
			else{
				continu--;
			}	
		}
	}

        public void getBestSDupbound(double mr){
                //System.out.println(mindiatan+"-"+maxdistan);
                double SD=0;
                int continu=5;
                double avgrArray[] = getrArray(mr);
                for(int r=0 ; r<avgrArray.length && continu>0 ; r++){
                        double avg=0;
                        double sum1=0;
                        double sum2=0;
                        double newSD = 0;
                        int tempNeborNum[] = new int[atomArray.length];
                        getNeborNum(avgrArray[r], tempNeborNum);
                        //compute SD
                        for(int i=0 ; i<tempNeborNum.length ; i++){
                                sum1 += tempNeborNum[i];
                                sum2 += Math.pow(tempNeborNum[i], 2);
                        }
                        avg = sum1 / tempNeborNum.length;
                        newSD = Math.sqrt( (sum2/tempNeborNum.length)-Math.pow(avg, 2) );
			//System.out.println(" " + newSD + " " + avgrArray[r] + " " + mr + " " + continu);
                        //get better neborNum if SD is better
                        if (newSD > SD){
                                SD = newSD;
				continu = 5;
                                for(int i=0 ; i<atomArray.length ; i++)
                                        atomArray[i].neborNum = tempNeborNum[i];
                                this.cr = avgrArray[r];
				if(this.cr > mr){
					//this.cr = mr;
					getlrNeborNum(mr);
					continu=0;
				}
                        }
                        else{
                                continu--;
                        }
                }
        }

        private double[] getrArray(double kh){
                int h = (int) (kh+6);
                double avgrArray[] = new double[h*2];
                for(int t=0; t<h*2 ; t++){
                	avgrArray[t] = t*0.5 + 3;
			//System.out.println(t);
                }
                return avgrArray;
        }
	
	private double[] getAvgrArray(double kl, double kh){
		int l = (int) (atomdistanArray[0].length/kl);
		int h = (int) (atomdistanArray[0].length/kh);
		double avgrArray[] = new double[h-l];
		for(int j=l, t=0; j<h ; j++, t++){
			for(int i=0 ; i<atomdistanArray.length ; i++){
				avgrArray[t]+=atomdistanArray[i][j];
			}
			avgrArray[t] = avgrArray[t] / (atomdistanArray.length-1);
		}
		return avgrArray;
	}
	
	public void getlrNeborNum(double cr){
		int tempNeborNum[] = new int[atomArray.length];
		//getNeborNum
		getNeborNum(cr, tempNeborNum);
		//output NeborNum in atomArray.neborNum;
		for(int i=0 ; i<atomArray.length ; i++)
			atomArray[i].neborNum = tempNeborNum[i];
	}
	
	private void getNeborNum(double cr, int tempArray[]){
		for(int i=0 ; i<tempArray.length ; i++){
			int j = Arrays.binarySearch(atomdistanArray[i], cr);
			j = Math.abs(j+1)-1;
			tempArray[i] = j;	
		}
	}
	
	private double getDistance (double x1, double y1, double z1, double x2, double y2, double z2){
		return Math.sqrt(Math.pow(x2-x1, 2)+Math.pow(y2-y1, 2)+Math.pow(z2-z1, 2));
	}

	//[0]=pdb, [1]=prl, [2]=prh, [3,4,5]=gx,y,z, [6]=box number, [7]=ligen radius, [8]=beta, [9]=alpha
	//[10]=rmodule, [11]=gbmod, [12]=mr, [13]=pushmode
	public static void main(String[] args) throws IOException, NumberFormatException, InterruptedException{
		new ConnectBoxes(args[0], 
						 Float.parseFloat(args[1]), 
						 Float.parseFloat(args[2]),
						 (Float.parseFloat(args[3])/2), 
						 (Float.parseFloat(args[4])/2), 
						 (Float.parseFloat(args[5])/2), 
						 Integer.parseInt(args[6]),
						 Float.parseFloat(args[7]),
						 Float.parseFloat(args[8]),
						 Float.parseFloat(args[9]), 
						 Integer.parseInt(args[10]),	//rmodule
						 Integer.parseInt(args[11]),	//get blue atom mod
						 Float.parseFloat(args[12]),	//manual radius
						 Integer.parseInt(args[13])		//push mod
						);
	}
}

File: reactionver/connectboxes.pl


sub connectboxes{ #The main procedure: to get [$boxnum] good boxes above receptors [$fln]
	my $filename = shift; 	#pdbqt file name
	my $prl = shift;	#neighbor percentile low
	my $prh = shift;	#neighbor percentile hight
	my $gx = shift;
	my $gy = shift;
	my $gz = shift;
	my $boxnum = shift;
	my $lr = shift;		#ligand radius
	my $alpha = shift;	
	my $beta = shift;
	my $rmodule = shift;	# 0=lr mod, 1=BestSd mod, 2=manual radius
	my $gbmod = shift;	# get prh nebor atom. 0=prh of full ball, 1=prh of all nebor of all atom
	my $mr = shift;		# manual radius
	my $pushmod = shift;	#pushmod 0=gamma% blueatmom, 1=number of puhmod blueatom 
	my @a = `java ConnectBoxes $filename $prl $prh $gx $gy $gz $boxnum $lr $beta $alpha $rmodule $gbmod $mr $pushmod`;
	chomp @a;
	my @x;
	my @y;
	my @z;
	my $g=$beta+1;
	$x[0]=$gx/2; $x[1]=$gx/2*-1;
	$y[0]=$gy/2; $y[1]=$gy/2*-1;
	$z[0]=$gz/2; $z[1]=$gz/2*-1;
	#my @bp; #blue point the atom inside the mol
	#my @sp; #silver point the atom on the surface
	#my @bkp; #black point the atom for candidate
	#my @fp; #first point

		$t = shift @a;
		#print "$t\n";
	my $bound = $#a+1;
	for(my $i=0 ; $i<$bound ; $i++){
		$t = shift @a;
		@ta = split(" ",$t);
		if($ta[0]=~/Surface/){
			#print "$ta[0]\n";
			last;
		}
		#if($t[0] == "null"){
		#	last;
		#}
		$bp[$i][0] = $ta[0];
		$bp[$i][1] = $ta[1];
		$bp[$i][2] = $ta[2];
	}
	$bound = $#a+1;
        for(my $i=0 ; $i<$bound ; $i++){
                $t = shift @a;
                #@ta = split(" ",$t);
                if($t=~/First/){
			#print "$ta[0]\n";
                        last;
                }
                #if($t[0] == "null"){
                #       last;
                #}
                $sp[$i][0] = $t;
                $sp[$i][1] = 0;  # 0=red 1=black 2=yellow 
        }
	$t = shift @a;
	@fp = split(" ",$t);   #get the first box
#print "$bp[0][0] $bp[0][1] $bp[0][2]\n";
#print "$sp[0][0] $sp[0][1] $sp[0][2]\n";
#print "$fp[0] $fp[1] $fp[2]\n";

############## initial	
	bomb($t,$g,$x[0],$x[1],$y[0],$y[1],$z[0],$z[1]);
	push(@ans,$t);
####################
	my $hasb=1;
	while($hasb==1){
		$hasb = neighbor_select($g,$x[0],$x[1],$y[0],$y[1],$z[0],$z[1]);
		if($hasb==0){
			$hasb = redcheck($g,$x[0],$x[1],$y[0],$y[1],$z[0],$z[1]);
		}
	}
	#foreach $str(@ans){
	#	print "$str \n";	
	#}
	#print "$#ans \n";
	#$a = shift ;
	print "@ans \n";
	pushans($pushmod, $lr, $alpha);
	
	print "@ans \n";
	my @result;
	foreach $i(0...$#ans){
		my @t=split(" ",$ans[$i]);
		$result[$i][0]=$t[0];  $result[$i][1]=$t[1];  $result[$i][2]=$t[2]; 
	}
#print "$#result";
#print "$result[0][0] $result[0][1] $result[0][2]";
	return @result;
}

sub bomb{
	my $bc = shift;
	my @x,@y,@z;
	my $dd = shift;
	$x[0]=shift; $x[1]=shift;
        $y[0]=shift; $y[1]=shift;
        $z[0]=shift; $z[1]=shift;

	my @ctr = split(" ",$bc);
	foreach $i(0...$#sp){
		@pt=split(" ",$sp[$i][0]);
		if(($sp[$i][1]!=2)&&(($pt[0]-$ctr[0]<$x[0])&&($pt[0]-$ctr[0]>$x[1]))&&(($pt[1]-$ctr[1]<$y[0])&&($pt[1]-$ctr[1]>$y[1]))&&(($pt[2]-$ctr[2]<$z[0])&&($pt[2]-$ctr[2]>$z[1])))
		{
			$sp[$i][1]=2;  #bombed 2
		}
	}
        $x[0]=$x[0]*$dd; $x[1]=$x[1]*$dd;
        $y[0]=$y[0]*$dd; $y[1]=$y[1]*$dd;
        $z[0]=$z[0]*$dd; $z[1]=$z[1]*$dd;

        foreach $i(0...$#sp){
                @pt=split(" ",$sp[$i][0]);
		if(($sp[$i][1]==0)&&(($pt[0]-$ctr[0]<$x[0])&&($pt[0]-$ctr[0]>$x[1]))&&(($pt[1]-$ctr[1]<$y[0])&&($pt[1]-$ctr[1]>$y[1]))&&(($pt[2]-$ctr[2]<$z[0])&&($pt[2]-$ctr[2]>$z[1])))
                {
                        $sp[$i][1]=1;  #new candidate 1
                }
        }
}




sub neighbor_select{

        my @x,@y,@z;
        my $dd = shift;
        $x[0]=shift; $x[1]=shift;
        $y[0]=shift; $y[1]=shift;
        $z[0]=shift; $z[1]=shift;
	my $max=-1;
	my $maxnum=-1;
	my @ctr;	
	my $haveb=0;

	foreach $i(0...$#sp){
		if($sp[$i][1]==1){
			my $count=0;
			@ctr = split(" ",$sp[$i][0]);
			foreach $j(0...$#sp){
				@pt=split(" ",$sp[$j][0]);
				if(($sp[$i][1]!=2)&&(($pt[0]-$ctr[0]<$x[0])&&($pt[0]-$ctr[0]>$x[1]))&&(($pt[1]-$ctr[1]<$y[0])&&($pt[1]-$ctr[1]>$y[1]))&&(($pt[2]-$ctr[2]<$z[0])&&($pt[2]-$ctr[2]>$z[1])))
               			{
					$count++;
		                }
			}
			if($count>$max){
				$max=$count;
				$maxnum=$i;
			}
		}
	}
	if($maxnum == -1){
		return 0;
	}
	else{
		push(@ans,$sp[$maxnum][0]);
		bomb($sp[$maxnum][0],$dd,$x[0],$x[1],$y[0],$y[1],$z[0],$z[1]);
		return 1;
	}


}

sub redcheck{

        my @x,@y,@z;
        my $dd = shift;
        $x[0]=shift; $x[1]=shift;
        $y[0]=shift; $y[1]=shift;
        $z[0]=shift; $z[1]=shift;
        my $max=-1;
        my $maxnum=-1;
        my @ctr;

        foreach $i(0...$#sp){
                if($sp[$i][1]==0){
                        my $count=0;
                        @ctr = split(" ",$sp[$i][0]);
                        foreach $j(0...$#sp){
                                @pt=split(" ",$sp[$j][0]);
                                if(($sp[$i][1]==0)&&(($pt[0]-$ctr[0]<$x[0])&&($pt[0]-$ctr[0]>$x[1]))&&(($pt[1]-$ctr[1]<$y[0])&&($pt[1]-$ctr[1]>$y[1]))&&(($pt[2]-$ctr[2]<$z[0])&&($pt[2]-$ctr[2]>$z[1])))
                                {
                                        $count++;
                                }
                        }
                        if($count>$max){
                                $max=$count;
                                $maxnum=$i;
                        }
                }
        }
        if($maxnum == -1){
                return 0;
        }
        else{
                push(@ans,$sp[$maxnum][0]);
                bomb($sp[$maxnum][0],$dd,$x[0],$x[1],$y[0],$y[1],$z[0],$z[1]);
                return 1;
        }






}


sub pushans{

        my $bn = shift;
        my $lr = shift;
        my $alpha = shift;
        my $l = $lr*$alpha;

        foreach $pt(0...$#ans){
                my @t = split(" ",$ans[$pt]);
                my $db = 0;
                my $min = 9999;
                my @sum;
				my $uv;
				$sum[0][0] = 0; $sum[0][1] = 0; $sum[0][2] = 0;
                for(my $i=0;$i<$bn;$i++){
						$min = 9999;
						foreach(@bp){
							my @tt = split(" ",$_);
							my $dt = (($t[0]-$tt[0])**2+($t[1]-$tt[1])**2+($t[2]-$tt[2])**2)**0.5;
							if( $dt<$min && $dt>$db ){
								$min = $dt;
								$sum[1][0] = $tt[0]; $sum[1][1] = $tt[1]; $sum[1][2] = $tt[2];
							}
						}
						$db = $min;
						$sum[0][0] = $sum[0][0] + $sum[1][0]; 
						$sum[0][1] = $sum[0][1] + $sum[1][1]; 
						$sum[0][2] = $sum[0][2] + $sum[1][2]; 
                }
				$sum[0][0] = $sum[0][0]/$bn; $sum[0][1] = $sum[0][1]/$bn; $sum[0][2] = $sum[0][2]/$bn;
				$sum[0][0] = $t[0]-$sum[0][0]; 
				$sum[0][1] = $t[1]-$sum[0][1]; 
				$sum[0][2] = $t[2]-$sum[0][2];
				$uv = ($sum[0][0]**2+$sum[0][1]**2+$sum[0][2]**2)**0.5;
				$sum[0][0] = $t[0]+($sum[0][0]/$uv)*$l; 
				$sum[0][1] = $t[1]+($sum[0][1]/$uv)*$l; 
				$sum[0][2] = $t[2]+($sum[0][2]/$uv)*$l;
				$ans[$pt] = "$sum[0][0] $sum[0][1] $sum[0][2]";
        }

}


















1


